MobiFlight/Joysticks/VKB/VKBDevice.cs

```diff
@@ -0,0 +1,192 @@
+                    Buttons.Add(Encoders[encdec.Key].DeviceInc);
+                }
+            }
+        }
+
+        private static byte GetEncoderIndex(JoystickInput input)
+        {
+            return (byte)((input.Id - 1000) / 10);
+        }
+        private static VKBEncoder.EncoderAction GetEncoderAction(JoystickInput input)
+        {
+            return (VKBEncoder.EncoderAction)(input.Id % 10);
+        }
+        protected override void EnumerateOutputDevices()
+        {
+            Definition?.Outputs?.ForEach(output => Lights.AddChannel(output));
+            base.EnumerateOutputDevices();
+        }
+
+        public override void SetOutputDeviceState(string name, byte state)
+        {
+            Lights.UpdateState(name, state);
+        }
+
+
+        public override void UpdateOutputDeviceStates()
+        {
+            var data = Lights.CreateMessage();
+            if (data[7] == 0) return; // Only send message if there are non-zero LEDs to be updated
+            try
+            {
+                SendData(data);
+            }
+            catch (System.IO.IOException)
+            {
+                base.OnDeviceRemoved();
+            }
+        }
+        private void OnHidReportReceived(object sender, System.EventArgs e)
+        {
+            var inputReceiver = sender as HidDeviceInputReceiver;
+            while (inputReceiver.TryRead(InputReportBuffer, 0, out _))
+            {
+                byte ReportId = InputReportBuffer[0];
+                if (ReportId != 0x08) // 0x08 = Monitoring channel / virtual bus
+                    continue;
+                byte MessageType = InputReportBuffer[1];
+                if (MessageType != 0x13) // 0x13 = Encoder status
+                    continue;
+                ParseEncoderReport(InputReportBuffer);
+            }
+        }
+        private void ParseEncoderReport(byte[] Report)
+        {
+            byte sequenceNo = Report[2];
+            if (((lastSeqNo + 1) & 0xFF) != sequenceNo)
+            {
+                Log.Instance.log("Some VKB encoder messages may have been missed", LogSeverity.Debug);
+                // Can easily happen if many updates are sent in quick succession (like when an encoder is spun fast)
+                // Not a problem since API reports absolute state, so the updates are just received on the next received message.
+            }
+            lastSeqNo = sequenceNo;
+            byte encoderCount = Report[3];
+            int maxEncoders = (Report.Length - 4) / 2;
+            if (encoderCount > maxEncoders)
+            {
+                Log.Instance.log($"Log message reports {encoderCount} encoders, but only has space for {maxEncoders}. Some encoders were ignored.", LogSeverity.Warn);
+                encoderCount = (byte)maxEncoders;
+                // Should not occur in most real-life scenarios, but it is theoretically possible to construct a device with more encoders than the report can handle.
```

Comment:
Same deal, the comment should go before the if statement to explain why the if is taking place.